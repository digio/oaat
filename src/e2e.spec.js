/* eslint-disable @getify/proper-arrows/where */
const path = require('path');
const spawn = require('spawn-command');

describe('amuck CLI', () => {
  describe('top-level help', () => {
    it('should be displayed when not enough arguments are passed', async () => {
      const result = await runCommand();
      expect(result).toMatchSnapshot();
    });

    it('should be displayed when --help is passed', async () => {
      const result = await runCommand('--help');
      expect(result).toMatchSnapshot();
    });
  });

  describe('lint', () => {
    it('should lint the spec file using the default config', async () => {
      const outputFile = 'fixtures/output/lint1.json';
      await runCommand(`lint ./fixtures/noExamples.json -o ./${outputFile}`);

      const result = require(`../${outputFile}`);
      expect(Object.keys(result.paths)).toMatchSnapshot('paths');
      expect(Object.keys(result.components.schemas)).toMatchSnapshot('schemas');
    });

    it('should update the x-examples into the parameters and requestBody, using the default config', async () => {
      const outputFile = 'fixtures/output/lint2.json';
      const output = await runCommand(`lint ./fixtures/lintExamples.json -o ./${outputFile}`);
      console.log(output);
      const result = require(`../${outputFile}`);
      expect(Object.keys(result.paths)).toMatchSnapshot('paths');
    });
  });

  describe('record', () => {
    it('should update the response files using the default config, using noExamples.json', async () => {
      const result = await runCommand(`record ./fixtures/noExamples.json -d`);
      expect(result).toMatchSnapshot();
    });

    it('should update the response files using the default config, using threeExamples.json', async () => {
      const result = await runCommand(`record ./fixtures/threeExamples.json -d`);
      expect(result).toMatchSnapshot();
    });
  });

  describe('build', () => {
    it('should create a spec file that is OpenAPI 3.x compliant', async () => {
      const result = await runCommand(`record ./fixtures/noExamples.json -d`);
      expect(result).toMatchSnapshot();
    });
  });
});

function runCommand(args = '', cwd = process.cwd()) {
  const CLI_PATH = require.resolve('./cli');
  const isRelative = cwd[0] !== '/';

  if (isRelative) {
    cwd = path.resolve(__dirname, cwd);
  }

  return new Promise((resolve, reject) => {
    let stdout = '';
    let stderr = '';
    const command = `${CLI_PATH} ${args}`;
    const child = spawn(command, { cwd });

    child.on('error', (error) => {
      reject(error);
    });

    child.stdout.on('data', (data) => {
      stdout += data.toString();
    });

    child.stderr.on('data', (data) => {
      stderr += data.toString();
    });

    child.on('close', () => {
      if (stderr) {
        reject(stderr);
      } else {
        resolve(stdout);
      }
    });
  });
}
