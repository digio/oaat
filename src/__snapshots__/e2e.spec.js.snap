// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`amuck CLI lint should lint the spec file using the default config: paths 1`] = `
Array [
  "/albums",
  "/albums/{id}",
  "/albums/{id}/photos",
  "/comments",
  "/comments/{id}",
  "/photos",
  "/photos/{id}",
  "/posts",
  "/posts/{id}",
  "/posts/{id}/comments",
  "/todos",
  "/todos/{id}",
  "/users",
  "/users/{id}",
]
`;

exports[`amuck CLI lint should lint the spec file using the default config: schemas 1`] = `
Array [
  "Album",
  "Comment",
  "NotFoundError",
  "Photo",
  "Post",
  "Todo",
  "User",
]
`;

exports[`amuck CLI lint should update the x-examples into the parameters and requestBody, using the default config: paths 1`] = `
Array [
  "/posts",
  "/posts/{id}",
]
`;

exports[`amuck CLI record should update the response files using the default config, using noExamples.json 1`] = `
"[33mwarn[39m: Ignore (no x-examples parameters) - 200 get /posts
[33mwarn[39m: Ignore (no x-examples parameters) - 200 get /posts/{id}
[33mwarn[39m: Ignore (no x-examples parameters) - 404 get /posts/{id}
[33mwarn[39m: Ignore (no x-examples parameters) - 200 get /posts/{id}/comments
[33mwarn[39m: Ignore (no x-examples parameters) - 404 get /posts/{id}/comments
[33mwarn[39m: Ignore (no x-examples parameters) - 200 get /comments
[33mwarn[39m: Ignore (no x-examples parameters) - 200 get /comments/{id}
[33mwarn[39m: Ignore (no x-examples parameters) - 404 get /comments/{id}
[33mwarn[39m: Ignore (no x-examples parameters) - 200 get /albums
[33mwarn[39m: Ignore (no x-examples parameters) - 200 get /albums/{id}
[33mwarn[39m: Ignore (no x-examples parameters) - 404 get /albums/{id}
[33mwarn[39m: Ignore (no x-examples parameters) - 200 get /albums/{id}/photos
[33mwarn[39m: Ignore (no x-examples parameters) - 404 get /albums/{id}/photos
[33mwarn[39m: Ignore (no x-examples parameters) - 200 get /photos
[33mwarn[39m: Ignore (no x-examples parameters) - 200 get /photos/{id}
[33mwarn[39m: Ignore (no x-examples parameters) - 404 get /photos/{id}
[33mwarn[39m: Ignore (no x-examples parameters) - 200 get /todos
[33mwarn[39m: Ignore (no x-examples parameters) - 200 get /todos/{id}
[33mwarn[39m: Ignore (no x-examples parameters) - 404 get /todos/{id}
[33mwarn[39m: Ignore (no x-examples parameters) - 200 get /users
[33mwarn[39m: Ignore (no x-examples parameters) - 200 get /users/{id}
[33mwarn[39m: Ignore (no x-examples parameters) - 404 get /users/{id}
[32minfo[39m: Making 0 requests...
"
`;

exports[`amuck CLI record should update the response files using the default config, using threeExamples.json 1`] = `
"[32minfo[39m: Making 4 requests...
[32minfo[39m: Creating response file GET_posts-200_DEFAULT.json
[32minfo[39m: Creating response file GET_posts_{id}-200_DEFAULT_id_1.json
[32minfo[39m: Creating response file GET_posts_{id}-200_id_2.json
[32minfo[39m: Creating response file GET_posts_{id}-404_DEFAULT_badParam.json
"
`;

exports[`amuck CLI top-level help should be displayed when --help is passed 1`] = `
"Usage: amuck <command>

Options:
  -V, --version                            output the version number
  -h, --help                               display help for command

Commands:
  lint [options] <jsonFile>                Tidy the API Spec up a bit
  record [options] <jsonFile> [serverUrl]  Record the responses of API spec file endpoint requests (optionally use a different server to make requests)
  apig [options] <jsonFile> [serverUrl]    Adds custom headers & Swagger UI endpoint to allow deployment of spec file to AWS API Gateway
  help [command]                           display help for command
"
`;

exports[`amuck CLI top-level help should be displayed when not enough arguments are passed 1`] = `
"Usage: amuck <command>

Options:
  -V, --version                            output the version number
  -h, --help                               display help for command

Commands:
  lint [options] <jsonFile>                Tidy the API Spec up a bit
  record [options] <jsonFile> [serverUrl]  Record the responses of API spec file endpoint requests (optionally use a different server to make requests)
  apig [options] <jsonFile> [serverUrl]    Adds custom headers & Swagger UI endpoint to allow deployment of spec file to AWS API Gateway
  help [command]                           display help for command
"
`;
